### ISSUE TITLE
Make depedenciesMeta inject by default for workspace packages?

### ISSUE THREAD

**Issue by deokseong-kim-toss** [pnpm/pnpm#4965]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

**Describe the user story**

We maintain many packages with monorepo, but we often make mistake "missing peers"
 
- packages/a 
   - deps: workspace:b
- packages/b
   - peerDeps: something
   
expect: mssing peers 'something' in 'b' 
result: nothing

`dependenciesMeta.*.inject` option helps, but it is quite annoying write `inject` when every deps. It also cause another human error. 

**Describe the solution you'd like**

npm options like "use-inject-workspace = true" 

**Describe the drawbacks of your solution**

I don't fully understand why inject should be in `package.json` instead a pnpm option, `use-inject-workspace` makes `inject` not explicit. 

**Describe alternatives you've considered**

`use-inject-workspace = dependencies` 
`use-inject-workspace = dependencies,devDependencies` 

or 

add inject when add workspace as dependency?

Question: 
Sorry for question. I wonder why pnpm don't use inject for all dependencies. Does it have performance drawback?

**Comment by zkochan** [comment]
> I wonder why pnpm don't use inject for all dependencies. Does it have performance drawback?

The main reason is that when you use inject and you rebuild the injected dependency, you somehow need to resync that dependency.

**Comment by blowsie** [comment]
I would agree with this decision, if a workspace package specifies a peerDependency, it is simply ignored when you install it from another workspace package, without any error or warning.

**Comment by NullVoxPopuli** [comment]
can the resync somehow happen automatically?

I'm using dependenciesMeta.injected right now to help out with a deep peer dependency, and my library that forwards that peer needs to be built, and those built assets aren't present in the injected version until I run install again.

Since this will be a common scenario for me (in a monorepo), I'd like this to somehow be more ergonomic

**Comment by zkochan** [comment]
Currently the packages will be automatically synced if they are built via a prepare script. But they are not resynced when someone rebuilds them. Probably this should be fixed.

**Comment by NullVoxPopuli** [comment]
anyone have ideas for how to re-sync dependencies when using turborepo? I could prefix all my scripts with `pnpm install`, but that'd only work once, because without changes detected, turbo would skip the prefixed pnpm install

**Comment by zkochan** [comment]
> only symlink those files entries (+ package.json of course)

This will not work with symlinks.

This will only work with hard links (that we use now) or copies. Symlinks are ignored when node.js resolves dependencies. So even though the symlink to project-b will be inside project-a's node_modules, the dependencies of project-b will be resolved from the node_modules of project-b because that is where the symlink points.

**Comment by NullVoxPopuli** [comment]
would the same idea work with hard links? instead of file-based hard-links, hard-link folders, based on `package.json#files` / dirname(`package.json#exports`)?

Do hardlinks still work if say, the whole `dist` folder is deleted, and re-generated by rollup?

**Comment by zkochan** [comment]
No, hard links are only possible upon files.

**Comment by NullVoxPopuli** [comment]
Has node rejected the idea of resolving symlinks?
Maybe we can change node.

**Comment by zkochan** [comment]
Node.js should resolve symlinks to their real locations. Otherwise, node.js would not work with pnpm.

**Comment by NullVoxPopuli** [comment]
How do you mean, you said:

> Symlinks are ignored when node.js resolves dependencies.

**Comment by zkochan** [comment]
For example, let's consider these two projects:

```
a
  index.js
  node_modules
    react (v17)

b
  index.js
  node_modules
    a --> ../../a
    react (v16)
```

If `b/index.js` will require `a`, `a` will resolve `react` from `a/node_modules/react`, not from `b/node_modules/react`. Because the real location of `a` is not inside `b/node_modules/a` but inside `a`. Node.js is looking for dependencies in the. real location.

**Comment by NullVoxPopuli** [comment]
does this behavior apply to all files or just the package root?
maybe if b's a is a real folder, with hard-link to a's package.json, and then symlinking a `dist` folder for the other contents? 

I think I need to test this out in a monorepo. I'll report back with that

**Comment by NullVoxPopuli** [comment]
I got distracted, but in the mean time, @simonihmig made a util that would make re-syncing easier -- I commented about maybe adding it to pnpm here: https://github.com/pnpm/pnpm/issues/6088#issuecomment-1433563254

**Comment by NullVoxPopuli** [comment]
Added a README to this project: https://github.com/NullVoxPopuli/pnpm-sync-dependencies-meta-injected

it at least allows folks to regain some productivity :sweat_smile: 
_and_ has a watch mode

**Comment by NullVoxPopuli** [comment]
@zkochan would it make sense to internalize this library as a daemon, or... how should we proceed?

**Comment by zkochan** [comment]
What do you mean by initializing it as a daemon? You mean adding it to pnpm as a new command?

Unfortunately, I cannot really dogfood this as I only injected mode with [Bit](https://bit.dev), which automatically writes to the "injected" dists when compiling. For now, if your package works, you can transfer it to the pnpm org and we can reference it from the docs.

**Comment by NullVoxPopuli** [comment]
> What do you mean by initializing it as a daemon? 

without build time integration, _something_ needs to be running to watch output directories (`package.json#files`, for example), and if those change, re-sync them to the injected folders.

> you can transfer it to the pnpm org and we can reference it from the docs.

sounds good, I'll make a demo tomorrow of with and without the tool to show how it works and add that to the readme before transfer

**Comment by NullVoxPopuli** [comment]
@zkochan -- how is transferring supposed to work? should I transfer it to your user account first?

![image](https://github.com/pnpm/pnpm/assets/199018/46c23b66-74e9-4cbd-b0d0-409a3fd0325a)
